### PR TITLE
Flex Attention HOP: Add support for flex decoding

### PR DESCRIPTION
**Pull request by joydddd**
# Flex Decoding
tl;dr This PR adds `flex_decoding` kernel to higher-order-op: `flex_attention` as the backend for multi-head attention decoding. 

Higher-order-op `flex_attention` was introduced in (https://github.com/pytorch/pytorch/pull/121845) to accept a user defined score modification callable (`score_mod`) and through `torch.compile`to create an efficient fused flash attention kernel instatiation. The `flex_attention` kernel is efficient for long queries (>512 tokens) attention. This PR introduces `flex_decoding` kernel as an alternative backend for `flex_attention` HOP to handle LLM inference where short queries (<32 tokens) attends to long key/value sequences.

### Details

LLM decoding iteratively attends each newly generated token ( query length = 1 ) to a long key/value context (up to 132k). `flex_attention` kernel only parallelizes attention along query length (M), batch size (B) and number of heads (H) dimension. LLM decoding lacks enough parallelism in the M dimension to fill up all SMs on the modern GPUs. 

`flex_decoding` adds parallelization along key/value sequence length (N). The key/value cache of a single head are split into multiple blocks and the query tokens attends to them in parallel. The results for the same head are then reduced across KV blocks to generate a global output. 

## Examples

Consider a Group Query Attention (GQA) decoding case, where a query token of 16 query heads (Hq) attends to 2 kv head (Hkv). Assume a batch size of 2 (B=2) and kv cache length of 4096 (N=4096). The attention kernel iteratively attends to newly generated query token (Mq = 1). 

We transform this problem into a Multiheaded Attention (MHA) problem by assuming a query length equal to number of query heads per kv heads, i.e. M=Hq//Hkv. 
The inputs to `flex_attention` HOP is thus a query of shape (B=2, H=Hkv=2, M=Hq//Hkv=8, D=64), key,value of shape (B=2, H=Hkv=2, N=4096, D=64, which lead to an intermediate attention score matrix of shape (2, 2, 8, 4096) and an output of shape (2, 2, 8, 64). 

```Python
import torch
from torch.nn.attention._flex_attention import _flex_attention as flex_attention

torch.manual_seed(0)

# Lets create some input tensors
# query of shape (B, Hkv, Hq//Hkv, D) 
# key/value of shape (B, Hkv, N, D) 
query = torch.randn(2, 2, 8, 64, device="cuda", dtype=torch.float32)
key = torch.randn(2, 2, 4096, 64, device="cuda", dtype=torch.float32)
value = torch.randn(2, 2, 4096, 64, device="cuda", dtype=torch.float32)

# Lets create a new score_modification checkerboard. 
def checkerboard(score, batch, head, token_q, token_kv):
    score = torch.where(torch.abs(token_kv - token_q) == 1, score * 0.5, score)
    score = torch.where(torch.abs(token_kv - token_q) == 2, score * 2.0, score)
    return score

# Lets call flex_attention with this new score modification for decoding. 
# The flex_attention HOP will chose flex_decoding as its backend since our query length (M) is only 8. 
output = flex_attention(query, key, value, score_mod=checkerboard)

compiled_flex_attention = torch.compile(flex_attention)
out_compiled = compiled_flex_attention (query, key, value, score_mod=checkerboard)

torch.testing.assert_close(output, out_compiled, atol=2e-2, rtol=2e-2)
```

## Future Plans 
- This PR does not implement load mask for score_mod function. This means if the score_mod functions takes a captured buffer along the M dimension , it must be padded to q length of 16, or next 2^n of query length if q_len > 16. 
i.e.
```python
q_scale = torch.randn(Hq//Hkv, device="cuda")
q_scale = torch.nn.functional.pad(q_scale, (0, 16-Hq//Hkv)) # Pad captured buffer
def bias_mod(score, batch, head, q, kv):
    score = score + q_scale[token_q]
    return score
```
- Backward path for short queries (<128 token) currently does not work because the `flex_attention_backward` kernel is lacking mask support and only takes query length of a multiple of 128. 
- Dynamic shape and max_autotuning is currently not working
- Add block sparse mask support (#129216 is a draft for flex_attention kernel) 
- Add explicit GQA support. (#130076 is a draft for GQA support on flex_attention kernel) 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by joydddd** [comment]
Needs Triton Fix https://github.com/triton-lang/triton/pull/4061

**Comment by joydddd** [comment]
Use inductor generated fused reduction kernel. Performance is similar to hand written templated kernel.

**Comment by joydddd** [comment]
Split_kv tuning: one block per SM. 


### Speedups w.r.t Xformers SplitK Kernel
~1 -> avg. 1.22x max 1.8x

| Type    |   Speedup | score_mod   | dtype         | shape(B,Hkv,Hq//Hkv,M,D,N)   |   BandWidth (TB/s) |   GFlops/s |
|---------|-----------|-------------|---------------|------------------------------|--------------------|------------|
| Average |     1.224 |             |               |                              |                    |            |
| Max     |     1.846 | noop        | torch.float16 | (2, 1, 16, 64, 131072)       |              0.983 |  17425.572 |
| Min     |     0.998 | noop        | torch.float16 | (16, 2, 8, 64, 8192)         |              0.953 |   8438.743 |

File: test/inductor/test_flex_attention.py [hunk 1]
```diff
@@ -593,19 +633,63 @@ def func(q, k, v, k2, v2):
...
+    @common_utils.parametrize("v_s", test_input_strides)
+    @common_utils.parametrize("head_dims", test_Hq_Hkv)
+    def test_strided_decoding(self, dtype: torch.dtype, k_s, v_s, head_dims):
+        Hq, Hkv = head_dims
+        assert Hq % Hkv == 0
+        q1 = torch.randn((B * Hq * D), dtype=dtype, device="cuda")
+        k1 = torch.randn((B * Hkv * S * D * 4), dtype=dtype, device="cuda")
+        v1 = torch.randn((B * Hkv * S * D * 4), dtype=dtype, device="cuda")
+
+        q_shape = (B, H, Hq // Hkv, D)
```
Comment (drisspg) [inline test/inductor/test_flex_attention.py:650]: It might be more clear if (B, Hkv, Hq // hkv, D)

Comment (drisspg) [reply test/inductor/test_flex_attention.py:650]: or maybe name some things, group_size

also since users would be expected to do this GQA like grouping, we might want to add a utility for doing this.

File: test/inductor/test_flex_attention.py [hunk 2]
```diff
@@ -614,32 +698,37 @@ def score_mod_2(score, b, h, m, n):
...
-        head_scale = torch.randn(H, device="cuda")
+    @common_utils.parametrize("decoding", [False, True])
+    def test_captured_buffers_all_dims(self, dtype: torch.dtype, decoding):
+        head_scale = torch.randn(Hkv if decoding else H, device="cuda")
         batch_scale = torch.randn(B, device="cuda")
-        tok_scale = torch.randn(S, device="cuda")
+        kv_scale = torch.randn(S, device="cuda")
+        q_scale = torch.randn(Hq // Hkv if decoding else S, device="cuda")
+        if q_scale.shape[-1] < 16:
+            q_scale = torch.nn.functional.pad(q_scale, (0, 16 - Hq // Hkv))
```
Comment (drisspg) [inline test/inductor/test_flex_attention.py:723]: If users would need to pad the buffers they are going to capture, maybe we should handle this during the lowering? this feels hard to to use

File: test/inductor/test_flex_attention.py (hunk 1 above)
Comment (joydddd) [reply test/inductor/test_flex_attention.py:650]: sorry, my bad. This q_shape isn't used in the code below.(B, Hkv, Hq//Hkv, D) is the shape actually used.

File: torch/_inductor/kernel/flex_decoding.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+    # initialize pointer to m and l
+    m_i = tl.zeros([BLOCK_M], dtype=tl.float32) - float("inf")
+    l_i = tl.zeros([BLOCK_M], dtype=tl.float32)
+    acc = tl.zeros([BLOCK_M, BLOCK_DMODEL], dtype=tl.float32)
+
+    if SAFE_M_BOUNDARY:
+        q = tl.load(Q_block_ptr)
+    else:
+        q = tl.load(Q_block_ptr, boundary_check=(0, ))
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_decoding.py:155]: Does this impact perf?

File: torch/_inductor/kernel/flex_decoding.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+        tl.store(L_block_ptr, l_i[None, :], boundary_check=(1,))
+
+    # -- store output
+    idx_z = off_z
+    idx_h = off_h
+    idx_t = off_t
+    idx_m = offs_m[:, None]
+    idx_d = offs_d[None, :]
+    # TODO generalize and add proper mask support
+    mask = (idx_m < Q_CTX) & (idx_d != -1)
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_decoding.py:227]: I think you only need `mask = (idx_m < Q_CTX)`.

File: torch/_inductor/kernel/flex_decoding.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    return split_k
+
+
+def _get_decoding_default_config(key) -> Tuple[int, int, int]:
+    B = key.get_size()[0]
+    H = key.get_size()[1]
+    Mk = key.get_size()[-2]
+    head_dim = key.get_size()[-1]
+    dtype = key.get_dtype()
+    default_config = None
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_decoding.py:251]: This function seems to havea  lot of dead code?

File: torch/_inductor/kernel/flex_decoding.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+    layout_acc = FixedLayout(
+        query.get_device(),
+        torch.float32,
+        buf_ACC_shape,
+        FlexibleLayout.contiguous_strides(buf_ACC_shape),
+    )
+
+    BLOCK_M = (
+        query.get_size()[-2]
+        if query.get_size()[-2] <= 0
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_decoding.py:305]: add a note that this is intentional.

**Review by Chillee — CHANGES_REQUESTED** [review]
I would prefer to have a separate `test_flex_decoding.py` test file for now, i think. There's too many changes and various workarounds for tests here.

In general, I think that I generally lean towards making tests easy to understand what they're testing (and be fine with duplication).

File: test/inductor/test_flex_attention.py [hunk 7]
```diff
@@ -547,7 +587,7 @@ def func(q, k, v):
         _, code = run_and_get_code(func, q, k, v)
         # Ensure _create_block_mask_from_mask is compiled and generates 3 kernels,
         # flex_attention generates 1 kernel.
-        FileCheck().check_count(".run(", 4, True).run(code[0])
+        FileCheck().check_count(".run(", 4, False).run(code[0])
```
[the remark this answers is not in the document]
Comment (joydddd) [reply test/inductor/test_flex_attention.py:590]: To allow lowering of mask generation generate different number of kernels. It was failing in trunk. We might need to a way to better test this.

File: test/inductor/test_flex_attention.py (hunk 2 above)
Comment (joydddd) [reply test/inductor/test_flex_attention.py:723]: The ultimate solution is adding load mask to modification{{}} generated code. Like this: 
https://github.com/joydddd/pytorch/pull/2/files# But we have some performance problems.  

Just a temp work around before we figure that out.

File: test/inductor/test_flex_attention.py [hunk 8]
```diff
@@ -217,7 +227,15 @@ def run_test(
         KV_H: int = H,
         KV_S: int = S,
         KV_D: int = D,
+        Hq: int = Hq,
```
[the remark this answers is not in the document]
Comment (joydddd) [reply test/inductor/test_flex_attention.py]: Flex_decoding in a new file now.

File: torch/_inductor/kernel/flex_decoding.py (hunk 6 above)
Comment (joydddd) [reply torch/_inductor/kernel/flex_decoding.py:306]: This is a works around before Triton fix https://github.com/triton-lang/triton/pull/4061 is in pin.

File: torch/_inductor/kernel/flex_attention.py [hunk 9]
```diff
@@ -370,6 +375,51 @@ def build_subgraph_buffer(
 )
 
 
+def _use_flex_decoding(query):
+    # Decide which kernel to use, return true if use flex decoding kernel.
+    return V.graph.sizevars.evaluate_expr(sympy.Lt(query.get_size()[-2], 32 + 1))
+
+
+def filtered_configs(
```
[the remark this answers is not in the document]
Comment (joydddd) [reply torch/_inductor/kernel/flex_attention.py:381]: I would like to handle decoding cases where q seq len = 64 by flex_attention. After some performance benchmarking, I think it makes sense to direct them to decoding as well. This filter is now removed. 

Eventually we probably would need a better way to decide between kernels?

**Comment by joydddd** [comment]
@pytorch-bot merge

**Review by Chillee — APPROVED** [review]
